### PR TITLE
Add assertions for GetHashCode

### LINKED ISSUE #58226: Change of behavior in symbol comparison with SymbolEqualityComparer
**Version Used**: 
`Microsoft.CodeAnalysis.CSharp` nuget package version `3.10.0` and `3.11.0`

**Steps to Reproduce**:
The following code behaves differently on version 3.10.0 and 3.11.0
```csharp
using System;
using System.Linq;
using System.Collections.Generic;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

var code = @"
using System.Linq;

M1(args => string.Join("" "", args.Select(a => a.ToString())));
void M1<TResult>(Func<object?[], TResult>? f) { }";

var compilation = CSharpCompilation.Create(null)
    .AddSyntaxTrees(SyntaxFactory.ParseSyntaxTree(code))
    .AddReferences(MetadataReference.CreateFromFile(typeof(object).Assembly.Location));

var tree = compilation.SyntaxTrees.Single();
var lambdaSyntax = tree.GetRoot().DescendantNodes().OfType<SimpleLambdaExpressionSyntax>().First();
var semanticModel1 = compilation.GetSemanticModel(tree);
var semanticModel2 = compilation.GetSemanticModel(tree);

var lambdaSymbol = (IMethodSymbol)semanticModel1.GetSymbolInfo(lambdaSyntax).Symbol;
var p1 = lambdaSymbol.Parameters.Single();

var p2 = semanticModel2.GetDeclaredSymbol(lambdaSyntax.Parameter);

Console.WriteLine(p1.Equals(p2));                                // 3.10.0: True  | 3.11.0: True
Console.WriteLine(p1 == p2);                                     // 3.10.0: False | 3.11.0: False
Console.WriteLine(SymbolEqualityComparer.Default.Equals(p1, p2));// 3.10.0: True  | 3.11.0: True

Console.WriteLine(p1.GetHashCode() == p2.GetHashCode());         // 3.10.0: True  | 3.11.0: False <- DIFFERS
var s = new HashSet<ISymbol>(SymbolEqualityComparer.Default);
s.Add(p1);
Console.WriteLine(s.Contains(p2));                               // 3.10.0: True  | 3.11.0: False <- DIFFERS
```

**Expected Behavior**:
I expected `s.Contains(p2)` to be `true` on the last line. This is also the suggested behavior by https://github.com/dotnet/roslyn-analyzers/issues/4568.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #58226

This is the same as #58247, but with few added Debug assertions in product code. Either this PR or #58247 needs to be merged.

### CODE REVIEW

**Comment by Youssef1313** [comment]
PRs are out of sync now. If we want the assertions, we can cherry-pick commits starting from `8696f62`

File: src/Compilers/CSharp/Portable/Symbols/PublicModel/Symbol.cs [hunk 1]
```diff
@@ -98,7 +99,9 @@ bool ISymbol.Equals(ISymbol other, CodeAnalysis.SymbolEqualityComparer equalityC
 
         protected bool Equals(Symbol other, CodeAnalysis.SymbolEqualityComparer equalityComparer)
         {
-            return other is object && UnderlyingSymbol.Equals(other.UnderlyingSymbol, equalityComparer.CompareKind);
+            var areEqual = other is object && UnderlyingSymbol.Equals(other.UnderlyingSymbol, equalityComparer.CompareKind);
+            Debug.Assert(!areEqual || this.GetHashCode() == other.GetHashCode(), "Hash code should be the same for equal symbols.");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/PublicModel/Symbol.cs:103]: >Debug.Assert(!areEqual || this.GetHashCode() == other.GetHashCode(), "Hash code should be the same for equal symbols."); [](http://example.com/codeflow?start=12&length=120)

Is this going to duplicate check performed by SymbolEqualityComparer?

**Comment by AlekseyTs** [comment]
>     internal sealed class SymbolEqualityComparer : EqualityComparer<Symbol>

Does VB have similar type?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/SymbolEqualityComparer.cs:13 in 131f146. [](commit_id = 131f14671e49ea48bf3fb7ecd20d647c8d548b0f, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/PublicModel/Symbol.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/PublicModel/Symbol.cs:103]: @AlekseyTs Yep. Some callers go into a single path of the two.

**Comment by Youssef1313** [comment]
> > ```
> > internal sealed class SymbolEqualityComparer : EqualityComparer<Symbol>
> > ```
> 
> Does VB have similar type?
> 
> Refers to: src/Compilers/CSharp/Portable/Symbols/SymbolEqualityComparer.cs:13 in [131f146](https://github.com/dotnet/roslyn/commit/131f14671e49ea48bf3fb7ecd20d647c8d548b0f). [](commit_id = [131f146](https://github.com/dotnet/roslyn/commit/131f14671e49ea48bf3fb7ecd20d647c8d548b0f), deletion_comment = False)

Nope. SymbolEqualityComparer was introduced with nullable reference types so that consumers can specify whether they want `ReferenceType` and `ReferenceType?` equal or not. So it's not a thing in VB.

**Comment by Youssef1313** [comment]
@AlekseyTs Let me know if you want me to re-open the PR with the assertion commits only.

**Comment by AlekseyTs** [comment]
> Let me know if you want me to re-open the PR with the assertion commits only.

Let's wait for the other PR to go in. I am curious how much this is going to slow down full test run.

**Pull request closed without merging**